src/commands/execution_ctx.c

```diff
@@ -15,74 +16,97 @@ static ExecutionType _GetExecutionTypeFromAST(AST *ast) {
-	if(exec_ctx) {
-		// Cache hit - Clone ast and plan. Set the execution type for query execution and indicate a cache hit.
-		*ast = AST_ShallowCopy(exec_ctx->ast);
-		// Set AST as it is retrived from cache.
-		QueryCtx_SetAST(exec_ctx->ast);
-		*plan = ExecutionPlan_Clone(exec_ctx->exec_plan_template);
-		exec_type = EXECUTION_TYPE_QUERY;
-		*cached = true;
+	ExecutionCtx *cached_exec_ctx = Cache_GetValue(cache, query_string);
+	if(cached_exec_ctx) {
+		// Cache hit - Clone the execution context. Set the execution type for query execution and indicate a cache hit.
+		// Set AST as it is retrived from cache and it is required for execution plan clone.
+		QueryCtx_SetAST(cached_exec_ctx->ast);
 		// Set parameters parse result in the execution ast.
-		AST_SetParamsParseResult(*ast, params_parse_result);
-		return exec_type;
+		AST_SetParamsParseResult(cached_exec_ctx->ast, params_parse_result);
 	}
-	// No cached execution plan, try to parse the query.
+	return cached_exec_ctx;
+}
+
+static AST  *_ExecutionCtx_ParseAST(const char *query_string,
+									cypher_parse_result_t *params_parse_result) {
 	cypher_parse_result_t *query_parse_result = parse_query(query_string);
 	// If no output from the parser, the query is not valid.
 	if(!query_parse_result) {
 		parse_result_free(params_parse_result);
-		return exec_type;
+		return NULL;
 	}
 
-	// Prepare the constructed AST for accesses from the module
-	AST *ast_template = AST_Build(query_parse_result);
-	exec_type = _GetExecutionTypeFromAST(ast_template);
-	*ast = ast_template;
+	// Prepare the constructed AST.
+	AST *ast = AST_Build(query_parse_result);
 	// Set parameters parse result in the execution ast.
-	AST_SetParamsParseResult(*ast, params_parse_result);
-	// In case of valid query, create execution plan, cache it, and return a clone.
+	AST_SetParamsParseResult(ast, params_parse_result);
+	return ast;
+}
+
+ExecutionCtx ExecutionCtx_FromQuery(const char *query) {
+	// Have an invalid ctx for errors.
+	ExecutionCtx invalid_ctx = {0};
+	const char *query_string;
+	// Parse and validate parameters only. Extract query string.
+	cypher_parse_result_t *params_parse_result = parse_params(query, &query_string);
+	// Return invalid execution context if there isn't a parser result.
+	if(params_parse_result == NULL) return invalid_ctx;
+
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	Cache *cache = GraphContext_GetCache(gc);
+	// Check the cache to see if we already have a cached context for this query.
+	ExecutionCtx *cached_exec_ctx = _ExecutionCtx_FromCache(cache, query_string, params_parse_result);
+	if(cached_exec_ctx) {
+		ExecutionCtx ctx = _ExecutionCtx_Clone(*cached_exec_ctx);
+		ctx.cached = true;
+		return ctx;
+	}
+
+	// No cached execution plan, try to parse the query.
+	AST *ast = _ExecutionCtx_ParseAST(query_string, params_parse_result);
+	// Invalid query, return invalid execution context.
+	if(!ast) return invalid_ctx;
```

Comment:
Don't we need to free `params_parse_result` ?

Reply:
taken care in 
```
static AST *_ExecutionCtx_ParseAST(const char *query_string,
								   cypher_parse_result_t *params_parse_result) {
	cypher_parse_result_t *query_parse_result = parse_query(query_string);
	// If no output from the parser, the query is not valid.
	if(!query_parse_result) {
		parse_result_free(params_parse_result);
		return NULL;
	}

	// Prepare the constructed AST.
	AST *ast = AST_Build(query_parse_result);
	// Set parameters parse result in the execution ast.
	AST_SetParamsParseResult(ast, params_parse_result);
	return ast;
}
```